### PR TITLE
Simple supplier stock management

### PR DESCRIPTION
**Pull request by tulimaki**
* Enable stock_behavior at product edit admin
* Limit suppliers to one as a default
* Add new features to picotable
* Update stock values when needed
* Enhance simple_supplier app to handle stock
* Enable checking orderability for products with supplier
* Add tests for simple_supplier app

Refs SHOOP-1876

### CODE REVIEW

File: shoop/core/models/_orders.py [hunk 1]
```diff
@@ -345,6 +345,9 @@ def save(self, *args, **kwargs):
         super(Order, self).save(*args, **kwargs)
         if first_save:  # Have to do a double save the first time around to be able to save identifiers
             self._save_identifiers()
+        for line in self.lines.all():
+            if line.product_id:
```
Comment (akx) [inline shoop/core/models/_orders.py]: Do this filtering during the queryset: `for line in self.lines.exclude(product_id=None)`

File: shoop/core/models/_shipments.py [hunk 2]
```diff
@@ -54,6 +54,11 @@ def __repr__(self):  # pragma: no cover
             self.pk, self.order_id, self.tracking_code, self.created_on
         )
 
+    def save(self, *args, **kwargs):
+        super(Shipment, self).save(*args, **kwargs)
+        for product in self.products.all():
```
Comment (akx) [inline shoop/core/models/_shipments.py]: Optimize:
```python
for product_id in self.products.values_list("id", flat=True):
   self...(product_id=product_id)
```

File: shoop/core/suppliers/base.py [hunk 3]
```diff
@@ -61,7 +61,7 @@ def get_orderability_errors(self, shop_product, quantity, customer):
         if stock_status.error:
             yield ValidationError(stock_status.error, code="stock_error")
         if shop_product.product.stock_behavior == StockBehavior.STOCKED:
-            if quantity >= stock_status.logical_count:
+            if quantity > stock_status.logical_count:
```
Comment (akx) [inline shoop/core/suppliers/base.py]: There are two hard things in computer science: naming things, cache invalidation and off-by-one errors.

File: shoop/simple_supplier/models.py [hunk 4]
```diff
@@ -27,6 +36,21 @@ class StockCount(models.Model):
...
+    def currency(self):
+        return SHOOP_HOME_CURRENCY
+
+    @property
+    def includes_tax(self):
+        return False
+
+    @property
+    def stock_unit_price_value(self):
+        return self.stock_value_value / self.logical_count if self.logical_count else 0
```
Comment (akx) [inline shoop/simple_supplier/models.py]: Parens, please

File: shoop/simple_supplier/admin_module/views.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+
+class StocksListView(PicotableListView):
+    template_name = "shoop/simple_supplier/admin/base_picotable.jinja"
+    model = Product
+    columns = [
+        Column(
+            "sku", _("SKU"), display="product__sku", linked=True,
+            filter_config=TextFilter(filter_field="sku", placeholder=_("Filter by sku..."))
```
Comment (akx) [inline shoop/simple_supplier/admin_module/views.py]: > Filter by SKU...

File: shoop/simple_supplier/static/simple_supplier/js/extra.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * This file is part of Shoop.
+ *
+ * Copyright (c) 2012-2016, Shoop Ltd. All rights reserved.
+ *
+ * This source code is licensed under the AGPLv3 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+function adjustStock(button) {
+    event.preventDefault();
+    var stockAdjustDiv= button.parent("div.form");
```
Comment (akx) [inline shoop/simple_supplier/static/simple_supplier/js/extra.js]: Doesn't look like this file has had eslint run on it.

File: shoop/simple_supplier/utils.py [hunk 7]
```diff
@@ -31,3 +49,68 @@ def get_current_stock_value(supplier_id, product_id):
...
+    :type supplier: shoop.core.models.Supplier
+    :param product: shoop Product
+    :type product: shoop.core.models.Product
+    :return: html div as a sting
+    :rtype: str
+    """
+    latest_adjustment = StockAdjustment.objects.filter(
+        product=product, supplier=supplier).order_by("-created_on").first()
+    context = {
+        "csrf_token": request.COOKIES.get("csrftoken"),
```
Comment (akx) [inline shoop/simple_supplier/utils.py]: This shouldn't be necessary and might not even work right; instead create a RequestContext or pass the `request` to `render_to_string´ and you can use the regular `{% csrf_token %}` tag.

File: shoop/simple_supplier/utils.py [hunk 8]
```diff
@@ -31,3 +49,68 @@ def get_current_stock_value(supplier_id, product_id):
...
+
+    Contains inputs for purchase_price_value and delta
+
+    :param request: HTTP request
+    :type request: django.http.HttpRequest
+    :param supplier: shoop Supplier
+    :type supplier: shoop.core.models.Supplier
+    :param product: shoop Product
+    :type product: shoop.core.models.Product
+    :return: html div as a sting
```
Comment (akx) [inline shoop/simple_supplier/utils.py]: The Police lead singer?

File: shoop/simple_supplier/utils.py [hunk 9]
```diff
@@ -31,3 +49,68 @@ def get_current_stock_value(supplier_id, product_id):
...
+    :rtype: str
+    """
+    latest_adjustment = StockAdjustment.objects.filter(
+        product=product, supplier=supplier).order_by("-created_on").first()
+    context = {
+        "csrf_token": request.COOKIES.get("csrftoken"),
+        "product": product,
+        "supplier": supplier,
+        "purchase_price_input_name": get_purchase_price_input_name(),
+        "purchase_price_label": "%s [%s]" % (_("Purchase price per unit"), SHOOP_HOME_CURRENCY),
```
Comment (akx) [inline shoop/simple_supplier/utils.py]: This should probably use the i18n api to localize the currency name (or at least fetch the currency symbol).

File: shoop/simple_supplier/utils.py [hunk 10]
```diff
@@ -31,3 +49,68 @@ def get_current_stock_value(supplier_id, product_id):
...
+        "purchase_price_value": latest_adjustment.purchase_price_value if latest_adjustment else Decimal("0.00"),
```
Comment (akx) [inline shoop/simple_supplier/utils.py]: Parens around the rvalue, please.

File: shoop/simple_supplier/utils.py [hunk 11]
```diff
@@ -31,3 +49,68 @@ def get_current_stock_value(supplier_id, product_id):
...
+        "delta_input_name": get_delta_input_name(),
+        "delta_label": _("Quantity"),
+        "delta_value": 0,
+        "delta_step": pow(0.1, product.sales_unit.decimals),
+    }
+    return render_to_string("shoop/simple_supplier/admin/add_stock_form.jinja", context)
+
+
+def get_purchase_price_input_name():
```
Comment (akx) [inline shoop/simple_supplier/utils.py]: These should just be `MODULE_LEVEL_CONSTANT`s

File: shoop/simple_supplier/templates/shoop/simple_supplier/admin/add_stock_form.jinja [hunk 12]
```diff
@@ -0,0 +1,18 @@
+<div class="form" data-url="/sa/adjust-stock/{{ supplier.id }}/{{ product.id }}/">
+    <input type="hidden" name="csrfmiddlewaretoken" value="{{ csrf_token }}">
```
Comment (akx) [inline shoop/simple_supplier/templates/shoop/simple_supplier/admin/add_stock_form.jinja]: So yeah, as mentioned elsewhere, use `{% csrf_token %}`

File: shoop/simple_supplier/templates/shoop/simple_supplier/admin/add_stock_form.jinja [hunk 13]
```diff
@@ -0,0 +1,18 @@
+<div class="form" data-url="/sa/adjust-stock/{{ supplier.id }}/{{ product.id }}/">
```
Comment (akx) [inline shoop/simple_supplier/templates/shoop/simple_supplier/admin/add_stock_form.jinja]: `data-url` here should use the `{{ url() }}` tag.

File: shoop/simple_supplier/templates/shoop/simple_supplier/admin/add_stock_form.jinja [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    <input type="hidden" name="csrfmiddlewaretoken" value="{{ csrf_token }}">
+    <div class="required-field">
```
Comment (akx) [inline shoop/simple_supplier/templates/shoop/simple_supplier/admin/add_stock_form.jinja]: Actually, why aren't these using the bs3 form renderer?

File: shoop/admin/modules/suppliers/views/edit.py [hunk 15]
```diff
@@ -29,6 +31,14 @@ class SupplierEditView(CreateOrUpdateView):
     template_name = "shoop/admin/suppliers/edit.jinja"
     context_object_name = "supplier"
 
+    def get_object(self, queryset=None):
+        obj = super(SupplierEditView, self).get_object(queryset)
+        if not settings.SHOOP_ENABLE_MULTIPLE_SUPPLIERS:
```
Comment (akx) [inline shoop/admin/modules/suppliers/views/edit.py]: Should this be refactored into a free function (along with the equivalent code for Shops) like

```python
def check_only_one_allowed(setting_name, obj):
    if getattr(settings, setting_name, True):
        return
    if not obj.pk and obj.__class__.objects.count() >= 1:
        raise Problem(_("Only one %(model)s permitted.") % {"model": obj._meta.verbose_name})
# ...
check_only_one_allowed("SHOOP_ENABLE_MULTIPLE_SUPPLIERS", obj)
# ...
check_only_one_allowed("SHOOP_ENABLE_MULTIPLE_SHOPS", obj)
```

File: shoop/admin/templates/shoop/admin/products/_edit_base_form.jinja [hunk 16]
```diff
@@ -15,12 +15,7 @@
     {{ bs3.field(product_form.barcode) }}
     {{ bs3.field(product_form.gtin) }}
     {{ bs3.field(product_form.category) }}
-    {# TODO: This should be enabled in near future. #}
-    {#  bs3.field(product_form.stock_behavior) #}
-    <div class="form-group">
-        <label class="control-label" for="id_stock">{% trans %}Stock{% endtrans %}</label>
-        <input class="form-control" id="id_stock" maxlength="40" name="base-stock" value="{% trans %}Unstocked{% endtrans %}" disabled="disabled" type="text">
-     </div>
+    {{  bs3.field(product_form.stock_behavior) }}
```
Comment (Pikkupomo) [inline shoop/admin/templates/shoop/admin/products/_edit_base_form.jinja]: extra space before `bs3`

File: shoop/admin/utils/views.py [hunk 17]
```diff
@@ -137,6 +138,13 @@ def get_create_or_change_title(request, instance, name_field=None):
     return _("Unnamed %s") % instance._meta.verbose_name
 
 
+def check_only_one_allowed(setting_name, obj):
```
Comment (Pikkupomo) [inline shoop/admin/utils/views.py]: this actually raises and checks, `check_and_raise_only_one_allowed`?

File: shoop/core/models/_order_lines.py [hunk 18]
```diff
@@ -118,7 +118,9 @@ def save(self, *args, **kwargs):
         if self.product_id and not self.supplier_id:
             raise ValidationError("Order line has product but no supplier")
 
-        return super(OrderLine, self).save(*args, **kwargs)
+        super(OrderLine, self).save(*args, **kwargs)
+        if self.product_id:
+            self.supplier.module.update_stock(self.product_id)
```
Comment (Pikkupomo) [inline shoop/core/models/_order_lines.py]: no longer returns anything. This intented?

File: shoop/core/models/_orders.py [hunk 19]
```diff
@@ -347,6 +347,8 @@ def save(self, *args, **kwargs):
         super(Order, self).save(*args, **kwargs)
         if first_save:  # Have to do a double save the first time around to be able to save identifiers
             self._save_identifiers()
+        for line in self.lines.exclude(product_id=None):
+            line.supplier.module.update_stock(line.product_id)
```
Comment (Pikkupomo) [inline shoop/core/models/_orders.py]: line can have only one supplier with one module?

Comment (Pikkupomo) [inline shoop/core/models/_orders.py]: also doesn't order line save handle the stock updating?

File: shoop_tests/simple_supplier/test_simple_supplier.py [hunk 20]
```diff
@@ -40,12 +46,83 @@ def test_simple_supplier(rf):
...
+
+    # Add stocked children for product
+    child_product = create_product("child-test-product", shop, supplier)
+    child_product.stock_behavior = StockBehavior.STOCKED
+    child_product.save()
+    child_product.link_to_parent(product)
+
+    frm = SimpleSupplierForm(product=product, request=request)
+    assert len(frm.products) == 1
+    assert frm.products[0] == child_product
```
Comment (Pikkupomo) [inline shoop_tests/simple_supplier/test_simple_supplier.py]: Add more comments on tests to make them easier to follow please :+1:

File: shoop/core/models/_order_lines.py (hunk 18 above)
Comment (tulimaki) [inline shoop/core/models/_order_lines.py]: Yes. As far as I know models save does not have to return anything. And if I remember correctly the return option was not even used anywhere in our code.

File: shoop/core/models/_orders.py (hunk 19 above)
Comment (tulimaki) [inline shoop/core/models/_orders.py]: I am pretty sure there is cases when order changes (like cancelled) and order lines is not touched and it has effects on stock values. And line can have only one supplier and supplier can have only one module.

**Pull request merged**